Fix scoping issue for box edit tool coordinates

ndmlny-qs opened the pull request:
This commit modifies the name of the x & y coordinates that use the box edit tool, as the original name collided with a few glyph types.

All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [ ] issues: fixes #13500 
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

ndmlny-qs committed:
Rename x & y coordinates

This commit modifies the name of the x & y coordinates that use the box
edit tool, as the original name collided with a few glyph types.

ndmlny-qs commented:
The changes all for all supported glyph types to add boxes using the box edit tool. The linking issue shows the issue, and the below video shows the fix in action.

[box-edit-fix.webm](https://github.com/bokeh/bokeh/assets/60523020/dbb2fcfd-a2ac-43d4-b2f5-27ba550c7ac5)

bryevdv commented:
Honestly a little surprised this doesn't cause a compile error. Solution LGTM but I think @mattpap is back tomorrow so I'll give him a little more time to chime in with any comments. Otherwise we'll want to get this in 3.3.1 as well.

bryevdv commented:
@ndmlny-qs maybe it's just a video artifact, the new hbar case in video seemed a little off (the drawn hbar did not seem to correspond to your cursor) can you confirm that it functions correctly in actual practice?

ndmlny-qs commented:
@bryevdv you are correct, the `hbar` and `vbar` cases do have an offset when drawing. I'll investigate further.

bryevdv commented:
@ndmlny-qs hbar and vbar are center-positioned along their "axis" so my first guess is that those coordinate computations were wrong to begin with, and need to be averages (e.g `(ry0 + ry1) / 2` for `hbar`)

mattpap commented on bokehjs/src/lib/models/tools/edit/box_edit_tool.ts:
```diff
@@ -59,73 +59,73 @@ export class BoxEditToolView extends EditToolView {
-    const [x0, x1] = renderer_view.coordinates.x_scale.r_invert(sx0, sx1)
-    const [y0, y1] = renderer_view.coordinates.y_scale.r_invert(sy0, sy1)
+    const [rx0, rx1] = renderer_view.coordinates.x_scale.r_invert(sx0, sx1)
+    const [ry0, ry1] = renderer_view.coordinates.y_scale.r_invert(sy0, sy1)
```
`r` in `r_invert` stands for "range", but the conversion is from screen to data coordinates, so it would make more sense to use `d` prefix.

mattpap commented on bokehjs/src/lib/models/tools/edit/box_edit_tool.ts:
```diff
@@ -59,73 +59,73 @@ export class BoxEditToolView extends EditToolView {
-    const [y0, y1] = renderer_view.coordinates.y_scale.r_invert(sy0, sy1)
+    const [rx0, rx1] = renderer_view.coordinates.x_scale.r_invert(sx0, sx1)
+    const [ry0, ry1] = renderer_view.coordinates.y_scale.r_invert(sy0, sy1)
     const fields = (() => {
```
This would have caught the problem at the type-level:
```suggestion
    const fields: {[key: string]: number} | null = (() => {
```

ndmlny-qs committed:
Address PR comments

This commit changes the name of `rx[x|1]` and `ry[0|1]` to `dx[0|1]` and
`dy[0|1]`. It also attempts to fix the cursor not aligning with the
drawing of a box in the `hbar` and `vbar` cases. It fixes it in one
direction, but not in the other, as I think we need the cursor
coordinates in order to create a check.

ndmlny-qs commented:
There is still an issue around the cursor being off going in the opposite direction of where you start to add the box in the `hbar` and `vbar` cases; see below. I'm not sure what the fix for this is, as the cursor position for the screen and data always have `[s|d]x0 < [s|d]x1` or `[s|d]y0 < [s|d]y1`.

[hbar-vbar.webm](https://github.com/bokeh/bokeh/assets/60523020/f1a5818c-06dc-4c00-912f-fb008f9889a6)

bryevdv commented:
@ndmlny-qs did you try averaging the "center" coordinate as I suggested earlier? I don't see that in the latest commit. I am also not sure about doubling the width/heigh fields I think it just needs the averaged center coordinate

ndmlny-qs commented:
I did, here is what averaging does `(d[x|y]0 + d[x|y]1)/2` for `[v|h]bar`.

[avg.webm](https://github.com/bokeh/bokeh/assets/60523020/eb50c757-a234-4c2f-baa6-8575e82012a6)

mattpap committed:
Fix HBar/VBar coordinates

mattpap committed:
Allow setting key modifiers in PlotActions

mattpap committed:
Deal with throttling threshold

mattpap committed:
Add unit regression tests

mattpap commented:
```
✗ Widgets
 └── should allow CategoricalSlider
images differ by 8px (0.02%)
```
I fixed this in PR #13497.

mattpap commented:
@ndmlny-qs, thanks.

bryevdv commented:
This no longer applies cleanly so I'll skip it for 3.3.1
